File: packages/hook/src/lib/git.ts
/**
 * Git helpers for detecting changed files.
 *
 * Used by exec and task commands for placeholder substitution
 * (`{{CHANGED_FILES}}`, `{{CHANGED_PACKAGES}}`) and skip-if-no-changes logic.
 */

import { createHash } from "node:crypto";
import type { RunOptions, RunResult } from "./proc";
import { runCommand as _defaultRunCommand } from "./proc";

// Module-level function reference — replaceable for testing without
// polluting the shared module registry via mock.module().
let _run: (opts: RunOptions) => Promise<RunResult> = _defaultRunCommand;

/** @internal For testing only. Override the runCommand implementation. */
export function _setRunCommand(fn: (opts: RunOptions) => Promise<RunResult>): void {
	_run = fn;
}

/** @internal For testing only. Reset to the real runCommand. */
export function _resetRunCommand(): void {
	_run = _defaultRunCommand;
}

/** Options for `getChangedFiles`. */
export type ChangedFilesOptions = {
	/** Working directory (repo root). */
	cwd?: string;
	/** Only consider staged files (for pre-commit checks). */
	stagedOnly?: boolean;
	/** Filter to files matching this extension (e.g., `.go`, `.ts`). */
	fileExt?: string;
};

/**
 * Get the list of changed files in the current repository.
 *
 * When `stagedOnly` is true, uses `git diff --cached --diff-filter=ACMR`
 * to return only added, copied, modified, or renamed files (excludes
 * deletions — deleted paths don't exist on disk).
 *
 * When `stagedOnly` is false, uses `git status --porcelain` to catch
 * staged, unstaged, and untracked files, filtering out deletions.
 */
export async function getChangedFiles(opts: ChangedFilesOptions = {}): Promise<string[]> {
	const { cwd = process.cwd(), stagedOnly = false, fileExt = "" } = opts;

	const command = stagedOnly
		? "git diff --cached --name-only --diff-filter=ACMR"
		: "git status --porcelain -uall" +
			" | grep -vE '^D.|^.D'" +
			" | cut -c4-" +
			" | sed 's/.* -> //'";

	const result = await _run({ command, cwd, timeout: 30 });
	if (result.exitCode !== 0) return [];

	let files = result.output
		.split("\n")
		.map((f) => f.trim())
		.map((f) => (f.startsWith('"') && f.endsWith('"') ? f.slice(1, -1) : f))
		.filter(Boolean);

	if (fileExt) {
		const ext = fileExt.startsWith(".") ? fileExt : `.${fileExt}`;
		files = files.filter((f) => f.endsWith(ext));
	}

	return files;
}

/**
 * Deduplicate parent directories from a list of file paths.
 * Useful for Go-style `./pkg/...` test targeting.
 */
export function getChangedPackages(files: string[]): string[] {
	const dirs = new Set<string>();
	for (const file of files) {
		const parts = file.split("/");
		parts.pop(); // remove filename
		const dir = parts.length === 0 ? "./" : `./${parts.join("/")}`;
		dirs.add(dir);
	}
	return [...dirs].sort();
}

/**
 * Shell-quote a single token for safe inclusion in `sh -c` commands.
 * Wraps in single quotes and escapes embedded single quotes.
 */
function shellQuote(s: string): string {
	return `'${s.replace(/'/g, "'\\''")}'`;
}

/**
 * Replace `{{CHANGED_FILES}}` and `{{CHANGED_PACKAGES}}` placeholders in a
 * command string with the actual values.
 *
 * Each file path and package path is shell-quoted to prevent command
 * injection from repository-controlled file names containing metacharacters.
 */
export function substitutePlaceholders(command: string, files: string[]): string {
	let result = command;
	if (result.includes("{{CHANGED_FILES}}")) {
		result = result.replace("{{CHANGED_FILES}}", files.map(shellQuote).join(" "));
	}
	if (result.includes("{{CHANGED_PACKAGES}}")) {
		const pkgs = getChangedPackages(files);
		result = result.replace("{{CHANGED_PACKAGES}}", pkgs.map(shellQuote).join(" "));
	}
	return result;
}

/**
 * Check whether there are uncommitted changes in the repository.
 * Uses `git status --porcelain` which covers staged, unstaged, and untracked files.
 */
export async function hasUncommittedChanges(cwd?: string): Promise<boolean> {
	const result = await _run({
		command: "git status --porcelain -uall",
		cwd: cwd ?? process.cwd(),
		timeout: 15,
	});
	return result.output.trim().length > 0;
}

/**
 * Check whether there are staged changes in the repository.
 * Useful for pre-commit checks where only staged files matter.
 */
export async function hasStagedChanges(cwd?: string): Promise<boolean> {
	const result = await _run({
		command: "git diff --cached --stat",
		cwd: cwd ?? process.cwd(),
		timeout: 15,
	});
	return result.output.trim().length > 0;
}

/** Options for `detectChanges`. */
export type DetectChangesOptions = {
	/** Working directory (repo root). */
	cwd: string;
	/** File extension filter (e.g., `.go`, `.ts`). */
	fileExt?: string;
	/** Only consider staged files. */
	staged?: boolean;
};

/**
 * Detect whether there are relevant changes in the repository.
 *
 * Used by exec and sync commands for skip-if-no-changes logic.
 */
export async function detectChanges(opts: DetectChangesOptions): Promise<boolean> {
	if (opts.fileExt) {
		const files = await getChangedFiles({
			cwd: opts.cwd,
			stagedOnly: opts.staged ?? false,
			fileExt: opts.fileExt,
		});
		return files.length > 0;
	}
	return opts.staged ? hasStagedChanges(opts.cwd) : hasUncommittedChanges(opts.cwd);
}

/**
 * Get the current HEAD commit SHA.
 *
 * Returns the full 40-character hex SHA, or an empty string if git is
 * unavailable or the directory is not a repository.
 */
export async function getHeadSha(cwd?: string): Promise<string> {
	try {
		const result = await _run({
			command: "git rev-parse HEAD",
			cwd: cwd ?? process.cwd(),
			timeout: 15,
		});
		return result.output.trim();
	} catch {
		return "";
	}
}

/** Options for `computeFingerprint`. */
export type FingerprintOptions = {
	/** Working directory (repo root). */
	cwd: string;
	/** Only consider staged changes (for pre-commit checks). */
	staged?: boolean;
	/** Scope diff to files matching this extension (e.g., `.go`, `.ts`). */
	fileExt?: string;
};

/**
 * Compute a composite fingerprint of the repository state: HEAD + working tree.
 *
 * Returns `sha256(HEAD_SHA + "\n" + diff_output)`. This captures both:
 *   1. The current commit (HEAD changes → fingerprint changes)
 *   2. Uncommitted modifications (any edit → fingerprint changes)
 *
 * When `staged` is true, uses `git diff --cached` instead of `git diff HEAD`.
 * When `fileExt` is provided, the diff is scoped to files matching that
 * extension via a pathspec filter.
 *
 * Used by:
 *   - state save/append — records a baseline fingerprint for session change
 *     detection (no scoping, full-repo diff).
 *   - exec run/check — records and validates sentinel fingerprints scoped
 *     to the exec's file extension and staged flag.
 *
 * Returns an empty string when git is unavailable (treated as "unknown").
 */
export async function computeFingerprint(opts: FingerprintOptions): Promise<string> {
	const head = await getHeadSha(opts.cwd);
	if (!head) return "";

	const parts = opts.staged ? ["git", "diff", "--cached"] : ["git", "diff", "HEAD"];

	if (opts.fileExt) {
		const ext = opts.fileExt.startsWith(".") ? opts.fileExt : `.${opts.fileExt}`;
		parts.push("--", `'*${ext}'`);
	}

	let diff: string;
	try {
		const result = await _run({
			command: parts.join(" "),
			cwd: opts.cwd,
			timeout: 30,
		});
		diff = result.output;
	} catch {
		return "";
	}

	return createHash("sha256").update(`${head}\n${diff}`).digest("hex");
}

File: internal/hook/git.go
package hook

import (
	"crypto/sha256"
	"fmt"
	"os/exec"
	"strings"
)

// getChangedFiles returns files that have changed in the repo.
// When stagedOnly is true, uses git diff --cached.
// When fileExt is set, filters to files matching that extension.
func getChangedFiles(cwd string, stagedOnly bool, fileExt string) []string {
	var cmd *exec.Cmd
	if stagedOnly {
		cmd = exec.Command("git", "diff", "--cached", "--name-only", "--diff-filter=ACMR") //nolint:gosec // args are not user-controlled
	} else {
		cmd = exec.Command("sh", "-c", //nolint:gosec // pipeline is a fixed string
			"git status --porcelain -uall | grep -vE '^D.|^.D' | cut -c4- | sed 's/.* -> //'")
	}
	cmd.Dir = cwd
	out, err := cmd.Output()
	if err != nil {
		return nil // git errors treated as no changes
	}

	lines := strings.Split(strings.TrimSpace(string(out)), "\n")
	var files []string
	for _, line := range lines {
		f := strings.TrimSpace(line)
		if f == "" {
			continue
		}
		// Remove surrounding quotes from git output
		if len(f) >= 2 && f[0] == '"' && f[len(f)-1] == '"' {
			f = f[1 : len(f)-1]
		}
		files = append(files, f)
	}

	if fileExt != "" {
		ext := fileExt
		if !strings.HasPrefix(ext, ".") {
			ext = "." + ext
		}
		var filtered []string
		for _, f := range files {
			if strings.HasSuffix(f, ext) {
				filtered = append(filtered, f)
			}
		}
		files = filtered
	}

	return files
}

// detectChanges checks whether there are relevant changes in the repo.
func detectChanges(cwd string, fileExt string, staged bool) (bool, error) {
	if fileExt != "" {
		files := getChangedFiles(cwd, staged, fileExt)
		return len(files) > 0, nil
	}

	if staged {
		return hasStagedChanges(cwd)
	}
	return hasUncommittedChanges(cwd)
}

func hasUncommittedChanges(cwd string) (bool, error) {
	cmd := exec.Command("git", "status", "--porcelain", "-uall")
	cmd.Dir = cwd
	out, err := cmd.Output()
	if err != nil {
		return false, nil
	}
	return strings.TrimSpace(string(out)) != "", nil
}

func hasStagedChanges(cwd string) (bool, error) {
	cmd := exec.Command("git", "diff", "--cached", "--stat")
	cmd.Dir = cwd
	out, err := cmd.Output()
	if err != nil {
		return false, nil
	}
	return strings.TrimSpace(string(out)) != "", nil
}

// getHeadSHA returns the current HEAD commit SHA.
func getHeadSHA(cwd string) string {
	cmd := exec.Command("git", "rev-parse", "HEAD")
	cmd.Dir = cwd
	out, err := cmd.Output()
	if err != nil {
		return ""
	}
	return strings.TrimSpace(string(out))
}

// computeFingerprint computes a SHA-256 fingerprint of HEAD + working tree diff.
func computeFingerprint(cwd string, staged bool, fileExt string) string {
	head := getHeadSHA(cwd)
	if head == "" {
		return ""
	}

	args := []string{"diff"}
	if staged {
		args = append(args, "--cached")
	} else {
		args = append(args, "HEAD")
	}

	if fileExt != "" {
		ext := fileExt
		if !strings.HasPrefix(ext, ".") {
			ext = "." + ext
		}
		args = append(args, "--", fmt.Sprintf("*%s", ext))
	}

	cmd := exec.Command("git", args...)
	cmd.Dir = cwd
	out, err := cmd.Output()
	if err != nil {
		return ""
	}

	h := sha256.Sum256([]byte(head + "\n" + string(out)))
	return fmt.Sprintf("%x", h)
}

// getChangedPackages deduplicates parent directories from file paths.
func getChangedPackages(files []string) []string {
	dirs := map[string]bool{}
	for _, f := range files {
		parts := strings.Split(f, "/")
		if len(parts) <= 1 {
			dirs["./"] = true
		} else {
			dir := "./" + strings.Join(parts[:len(parts)-1], "/")
			dirs[dir] = true
		}
	}
	result := make([]string, 0, len(dirs))
	for d := range dirs {
		result = append(result, d)
	}
	return result
}

// substitutePlaceholders replaces {{CHANGED_FILES}} and {{CHANGED_PACKAGES}}.
func substitutePlaceholders(command string, files []string) string {
	result := command
	if strings.Contains(result, "{{CHANGED_FILES}}") {
		quoted := make([]string, 0, len(files))
		for _, f := range files {
			quoted = append(quoted, shellQuoteWrap(f))
		}
		result = strings.Replace(result, "{{CHANGED_FILES}}", strings.Join(quoted, " "), 1)
	}
	if strings.Contains(result, "{{CHANGED_PACKAGES}}") {
		pkgs := getChangedPackages(files)
		quoted := make([]string, 0, len(pkgs))
		for _, p := range pkgs {
			quoted = append(quoted, shellQuoteWrap(p))
		}
		result = strings.Replace(result, "{{CHANGED_PACKAGES}}", strings.Join(quoted, " "), 1)
	}
	return result
}

// shellQuoteWrap wraps a string in single quotes for safe shell usage.
func shellQuoteWrap(s string) string {
	return "'" + strings.ReplaceAll(s, "'", "'\\''") + "'"
}
